runtime/README.md

```diff
@@ -692,107 +791,628 @@ Command Code: `0x434B_4558` ("CKEX")
+| auth_req_result | u32      | Authorized: `0xDEADC0DE` |
+|                 |          | Denied: `0x21523F21`     |
+
+## Mailbox commands: Cryptographic Mailbox (2.0)
+
+These commands are used by the [Cryptograhic Mailbox](#cryptographic-mailbox-commands-new-in-20) system.
+
+### CM_SHA_INIT
+
+This starts the computation of a SHA hash of data, which may be larger than a single mailbox command allows. It also supports additional algorithms.
+
+The sequence to use these are:
+* 1 `CM_SHA_INIT` command
+* 0 or more `CM_SHA_UPDATE` commands
+* 1 `CM_SHA_FINAL` command
+
+For each command, the context from the previous command's output must be passed as an input.
+
+Command Code: `0x434D_5349` ("CMSI")
+
+*Table: `CM_SHA_INIT` input arguments*
+
+| **Name**       | **Type** | **Description**                   |
+| -------------- | -------- | --------------------------------- |
+| chksum         | u32      |                                   |
+| hash algorithm | u32      | Enum.                             |
+|                |          | Value 0 = reserved                |
+|                |          | Value 1 = SHA2-256                |
+|                |          | Value 2 = SHA2-384                |
+|                |          | Value 3 = SHA2-512                |
+| data size      | u32      |                                   |
+| data           | u8[...]  | Data to hash                      |
+
+*Table: `CM_SHA_INIT` output arguments*
+| **Name**     | **Type** | **Description**                            |
+| ------------ | -------- | ------------------------------------------ |
+| chksum       | u32      |                                            |
+| fips_status  | u32      | FIPS approved or an error                  |
+| context size | u32      |                                            |
+| context      | u8[...]  | Passed to `CM_SHA_UPDATE` / `CM_SHA_FINAL` |
+
+*Table: `CM_SHA_INIT` / `CM_SHA_UPDATE` / `CM_SHA_FINAL` internal context*
+| **Name**          | **Type** | **Description** |
+| ----------------- | -------- | --------------- |
+| input buffer      | u8[128]  |                 |
+| intermediate hash | u8[64]   |                 |
+| length            | u64      |                 |
+| hash algorithm    | u32      |                 |
+
+### CM_SHA_UPDATE
+
+This continues a SHA computation started by `CM_SHA_INIT` or from another `CM_SHA_UPDATE`.
+
+The context MUST be passed in from `CM_SHA_INIT` or `CM_SHA_UPDATE`.
+
+Command Code: `0x434D_5355` ("CMSU")
+
+*Table: `CM_SHA_UPDATE` input arguments*
+| **Name**       | **Type** | **Description**                      |
+| -------------- | -------- | ------------------------------------ |
+| chksum         | u32      |                                      |
+| context size   | u32      | Size of the context.                 |
+| context        | u8[...]  | From `CM_SHA_INIT` / `CM_SHA_UPDATE` |
+| data size      | u32      |                                      |
```

Comment:
do we care that data size field may become an unaligned access depending on context size?

Reply:
I think we could define a fixed-size context struct, sized to hold the largest context data possible, and make sure it's aligned properly.

Reply:
I worry about a fixed-size struct, as it means that if we define a new command or want to redefine the context for one command, we break compatibility with all commands that use contexts. With the variable-sized contexts, we wouldn't need to break compatibility if we needed to add a field to any context.

But we can stipulate that all contexts have sizes that are multiples of 32 bits.